### DISCUSSION TITLE
Error to use UMAL

### DISCUSSION THREAD

**Discussion by Keisuke168** [neuralhydrology/neuralhydrology#114] [Q&A]
Hi, thanks for the package.

I am trying to experiment with the UMAL head. However, I am getting the following error in the validation loop: 
`"RuntimeError: input.size(-1) must be equal to input_size. Expected 3, got 2". `

While the training loop concatenates the inputs with tau through `UMALTrainer._pre_model_hook`, it seems that the error is occurring due to the dimensions being different as tau is not concatenated during validation. 

Since I still don't quite understand UMAL, I'm not sure how to resolve this. If you have any ideas about this error, please let me know.

**Comment by gauchm** [comment]
I think we'll need more information to figure out what's going on. Can you please share the full error stack trace and the config file you use for training?

**Reply by Keisuke168** [reply]
Thank you for your response. This is full error stack trace.


---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
     13 # train
---> 14 start_training(cfg)
     16 # evaluate

File [~/neuralhydrology/training/train.py:23], in start_training(cfg)
     21     raise ValueError(f"Unknown head {cfg.head}.")
     22 trainer.initialize_training()
---> 23 trainer.train_and_validate()

File [~/neuralhydrology/training/basetrainer.py:222], in BaseTrainer.train_and_validate(self)
    219     self._save_weights_and_optimizer(epoch)
    221 if (self.validator is not None) and (epoch % self.cfg.validate_every == 0):
--> 222     self.validator.evaluate(epoch=epoch,
    223                             save_results=self.cfg.save_validation_results,
    224                             metrics=self.cfg.metrics,
    225                             model=self.model,
    226                             experiment_logger=self.experiment_logger.valid())
    228     valid_metrics = self.experiment_logger.summarise()
    229     print_msg = f"Epoch {epoch} average validation loss: {valid_metrics['avg_total_loss']:.5f}"

File [~/neuralhydrology/evaluation/tester.py:211], in BaseTester.evaluate(self, epoch, save_results, metrics, model, experiment_logger)
    207         self.cached_datasets[basin] = ds
    209 loader = DataLoader(ds, batch_size=self.cfg.batch_size, num_workers=0, collate_fn=ds.collate_fn)
--> 211 y_hat, y, dates, all_losses = self._evaluate(model, loader, ds.frequencies)
    213 # log loss of this basin plus number of samples in the logger to compute epoch aggregates later
    214 if experiment_logger is not None:

File [~/neuralhydrology/evaluation/tester.py:425], in BaseTester._evaluate(self, model, loader, frequencies)
    423     if not key.startswith('date'):
    424         data[key] = data[key].to(self.device)
--> 425 predictions, loss = self._get_predictions_and_loss(model, data)
    427 for freq in frequencies:
    428     if predict_last_n[freq] == 0:

File [~/neuralhydrology/evaluation/tester.py:534], in UncertaintyTester._get_predictions_and_loss(self, model, data)
    533 def _get_predictions_and_loss(self, model: BaseModel, data: Dict[str, torch.Tensor]) -> Tuple[torch.Tensor, float]:
--> 534     outputs = model(data)
    535     _, all_losses = self.loss_obj(outputs, data)
    536     predictions = model.sample(data, self.cfg.n_samples)

File [~/anaconda3/envs/nh-pyviz/lib/python3.10/site-packages/torch/nn/modules/module.py:1501], in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

File [~/neuralhydrology/modelzoo/cudalstm.py:68], in CudaLSTM.forward(self, data)
     66 # possibly pass dynamic and static inputs through embedding layers, then concatenate them
     67 x_d = self.embedding_net(data)
---> 68 lstm_output, (h_n, c_n) = self.lstm(input=x_d)
     70 # reshape to [batch_size, seq, n_hiddens]
     71 lstm_output = lstm_output.transpose(0, 1)

File [~/anaconda3/envs/nh-pyviz/lib/python3.10/site-packages/torch/nn/modules/module.py:1501], in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

File [~/anaconda3/envs/nh-pyviz/lib/python3.10/site-packages/torch/nn/modules/rnn.py:810], in LSTM.forward(self, input, hx)
    806     # Each batch of the hidden state should match the input sequence that
    807     # the user believes he[/she]is passing in.
    808     hx = self.permute_hidden(hx, sorted_indices)
--> 810 self.check_forward_args(input, hx, batch_sizes)
    811 if batch_sizes is None:
    812     result = _VF.lstm(input, hx, self._flat_weights, self.bias, self.num_layers,
    813                       self.dropout, self.training, self.bidirectional, self.batch_first)

File [~/anaconda3/envs/nh-pyviz/lib/python3.10/site-packages/torch/nn/modules/rnn.py:730], in LSTM.check_forward_args(self, input, hidden, batch_sizes)
    725 def check_forward_args(self,  # type: ignore[override]
    726                        input: Tensor,
    727                        hidden: Tuple[Tensor, Tensor],
    728                        batch_sizes: Optional[Tensor],
    729                        ):
--> 730     self.check_input(input, batch_sizes)
    731     self.check_hidden_size(hidden[0], self.get_expected_hidden_size(input, batch_sizes),
    732                            'Expected hidden[0] size {}, got {}')
    733     self.check_hidden_size(hidden[1], self.get_expected_cell_size(input, batch_sizes),
    734                            'Expected hidden[1] size {}, got {}')

File [~/anaconda3/envs/nh-pyviz/lib/python3.10/site-packages/torch/nn/modules/rnn.py:218], in RNNBase.check_input(self, input, batch_sizes)
    214     raise RuntimeError(
    215         'input must have {} dimensions, got {}'.format(
    216             expected_input_dim, input.dim()))
    217 if self.input_size != input.size(-1):
--> 218     raise RuntimeError(
    219         'input.size(-1) must be equal to input_size. Expected {}, got {}'.format(
    220             self.input_size, input.size(-1)))

RuntimeError: input.size(-1) must be equal to input_size. Expected 3, got 2


This is the config file. (I couldn't upload the yml file, so I uploaded it as a txt file.) 

[umal.txt](https://github.com/neuralhydrology/neuralhydrology/files/11395149/umal.txt)

**Comment by kratzert** [comment]
Fixed in #119

**Comment by danklotz** [comment]
Hi @Keisuke168

This was indeed a bug that sneaked into the code over time. Thank you for telling us. As @kratzert says, it should be fixed with our newest update. 

Good luck and all the best.
